translator/internal/internal_to_oc.go

```diff
@@ -0,0 +1,430 @@
+// Copyright 2019 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	occommon "github.com/census-instrumentation/opencensus-proto/gen-go/agent/common/v1"
+	ocresource "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	octrace "github.com/census-instrumentation/opencensus-proto/gen-go/trace/v1"
+	"github.com/golang/protobuf/ptypes"
+
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/internal"
+	"github.com/open-telemetry/opentelemetry-collector/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	tracetranslator "github.com/open-telemetry/opentelemetry-collector/translator/trace"
+)
+
+const sourceFormat = "ot_internal_trace"
+
+var (
+	defaultProcessID = 0
+	emptyStatus      = data.SpanStatus{}
+)
+
+func internalToOC(td data.TraceData) []consumerdata.TraceData {
+	ocTraceData := consumerdata.TraceData{
+		SourceFormat: sourceFormat,
+	}
+
+	resourceSpansList := td.ResourceSpans()
+
+	ocResourceSpansList := make([]consumerdata.TraceData, 0, len(resourceSpansList))
+
+	for _, resourceSpans := range resourceSpansList {
+		ocTraceData.Node, ocTraceData.Resource = internalResourceToOC(resourceSpans.Resource())
+		ocSpans := make([]*octrace.Span, 0, td.SpanCount())
+		for _, span := range resourceSpans.Spans() {
+			ocSpans = append(ocSpans, spanToOC(span))
+		}
+		ocTraceData.Spans = ocSpans
+		ocResourceSpansList = append(ocResourceSpansList, ocTraceData)
+	}
+
+	return ocResourceSpansList
+}
+
+func internalResourceToOC(resource *data.Resource) (*occommon.Node, *ocresource.Resource) {
+	if resource == nil {
+		return nil, nil
+	}
+
+	attrs := resource.Attributes()
+
+	ocNode := occommon.Node{}
+	ocResource := ocresource.Resource{}
+
+	if len(attrs) == 0 {
+		return &ocNode, &ocResource
+	}
+
+	labels := make(map[string]string, len(attrs))
+	for key, attributeValue := range attrs {
+		val := attributeValueToString(attributeValue)
+
+		switch key {
+		case conventions.OCAttributeResourceType:
+			ocResource.Type = val
+		case conventions.AttributeServiceName:
+			if ocNode.ServiceInfo == nil {
+				ocNode.ServiceInfo = &occommon.ServiceInfo{}
+			}
+			ocNode.ServiceInfo.Name = val
+		case conventions.OCAttributeProcessStartTime:
+			t, err := time.Parse(time.RFC3339Nano, val)
+			if err != nil {
```

Comment:
You probably meant `if err == nil` here and below?
Also see https://github.com/open-telemetry/opentelemetry-collector/pull/614#discussion_r390688179, I suggest first discuss what is the preferred error handling and apply it both here and in OTLP translator.

Reply:
Also, it's better to use indentation for handling special cases like errors rather than for "happy path", i.e.
```
if err != nil {
  continue
}
...
``` 
It makes the code more readable and helps with avoiding typos like this one.

Reply:
See https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Reply:
Makes sense, thanks for catching it!